cqlconf.f90

```diff
@@ -0,0 +1,284 @@
+    character(len=8) :: cqlpmod
+    integer :: lrz
+    character(len=8) :: lrzdiff
+    integer :: lrzmax
+    integer :: lrindx(0:lrorsa)
+    integer :: ls
+    integer :: lsmax
+    character(len=8) :: lsdiff
+    integer :: lsindx(0:lrorsa)
+    character(len=8) :: nlrestrt
+    character(len=8) :: nlwritf
+
+    namelist/setup0/ mnemonic,ioutput,iuser,ibox,noplots,lnwidth, &
+         nmlstout,special_calls,cqlpmod,lrz,lrzdiff,lrzmax,lrindx, &
+         ls,lsmax,lsdiff,lsindx,nlrestrt,nlwritf
+
+    ! for now, we take thse from setup0,
+    ! because there is existing defaulting code earlier.
+    ! Ideally that code would live with the config setters and getters,
+    ! and this goes away.
+    ! To my knowledge fortran does not provide a better facility
+    mnemonic = setup0_%mnemonic
+    ioutput(2) = setup0_%ioutput(2)
+    iuser = setup0_%iuser
+    ibox(3) = setup0_%ibox(3)
+    noplots = setup0_%noplots
+    lnwidth = setup0_%lnwidth
+    nmlstout = setup0_%nmlstout
+    special_calls = setup0_%special_calls
+    cqlpmod = setup0_%cqlpmod
+    lrz = setup0_%lrz
+    lrzdiff = setup0_%lrzdiff
+    lrzmax = setup0_%lrzmax
+    lrindx(0:lrorsa) = setup0_%lrindx(0:lrorsa)
+    ls = setup0_%ls
+    lsmax = setup0_%lsmax
+    lsdiff = setup0_%lsdiff
+    lsindx(0:lrorsa) = setup0_%lsindx(0:lrorsa)
+    nlrestrt = setup0_%nlrestrt
+    nlwritf = setup0_%nlwritf
+
+
+    call maybe_nml_open(nml_file)
+    read(nml_fd, setup0)
+
+
+    ! external codes can call this
+    call set_setup0(mnemonic,ioutput,iuser,ibox,noplots,lnwidth, &
+         nmlstout,special_calls,cqlpmod,lrz,lrzdiff,lrzmax,lrindx, &
+         ls,lsmax,lsdiff,lsindx,nlrestrt,nlwritf)
+
+    if (present(close_nml_file)) then
+       if(close_nml_file .eqv. .TRUE.) then
+          call nml_close()
+       end if
+    endif
+
+  end subroutine get_setup0_from_nml
+
+  subroutine set_setup0(mnemonic,ioutput,iuser,ibox,noplots,lnwidth, &
+         nmlstout,special_calls,cqlpmod,lrz,lrzdiff,lrzmax,lrindx, &
+         ls,lsmax,lsdiff,lsindx,nlrestrt,nlwritf)
+    character(len=256), optional :: mnemonic
+    integer, optional :: ioutput(2)
+    character(len=8), optional :: iuser
+    character(len=8), optional :: ibox(3)
+    character(len=8), optional :: noplots
+    integer, optional :: lnwidth
+    character(len=8), optional :: nmlstout
+    character(len=8), optional :: special_calls
+    character(len=8), optional :: cqlpmod
+    integer, optional :: lrz
+    character(len=8), optional :: lrzdiff
+    integer, optional :: lrzmax
+    integer, optional :: lrindx(0:lrorsa)
+    integer, optional :: ls
+    integer, optional :: lsmax
+    character(len=8), optional :: lsdiff
+    integer, optional :: lsindx(0:lrorsa)
+    character(len=8), optional :: nlrestrt
+    character(len=8), optional :: nlwritf
+
+    ! Note, for all of these, we should collect old cql3d
+    ! default logic sprinkled around the code, and condense
+    ! it here
+
+    if (present(mnemonic)) then
```

Comment:
If we can get all the defaults moved to this module, we ditch the default branches in favor of a default instance of the type, which will be able to exist before the nml read, and also here.  Much nicer way to do it...

Reply:
done, much cleaner.